### ISSUE TITLE
Container initialization is very very slow

### ISSUE THREAD

**Issue by nicokaiser** [bitnami/bitnami-docker-wordpress#164]
**Description**

This is basically a duplicate of bitnami-docker-wordpress-nginx#6 but as it concerns the Bitnami WordPress package I move it to here.

**Steps to reproduce the issue:**

1. Clone the provided docker-compose.yml (as is)
2. docker-compose up
3. docker-compose down
4. docker-compose up

**Describe the results you received:**

The container startup is very slow. The `Initializing wordpress` step takes several minutes (!) to complete, every time the container is created.

**Describe the results you expected:**

I expect wordpress to be initialized once and then persisted to the volume (which is done: `wp-config.php` and `wp-contents` are stored on the volume, the `.initialized` file is generated).

```
wordpress_1  | nami    INFO  Initializing wordpress
wordpress_1  | wordpre INFO  WordPress has been already initialized, restoring...
```

Between these two lines there are multiple minutes of time wasted to initialize an already initialized instance.

**Additional information you deem important (e.g. issue happens only occasionally):**

When setting the log level to TRACE8, the log reveals that this might have something to to with nami trying to chown all files all over again (and doing so in a very slow way, as there are not that many files in the newly created volume). This should not happen when the `.initialized` file is found.

This also happens when using Docker Swarm and starting this container as a service. When restarting or updating (!) the service, it is offline for several minutes during the unnecessary initialization.

**Version**

- Output of `docker version`:

```
Client:
 Version:           18.09.3
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        774a1f4
 Built:             Thu Feb 28 06:34:04 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.3
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       774a1f4
  Built:            Thu Feb 28 05:59:55 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

- Output of `docker info`:

```
Containers: 27
 Running: 23
 Paused: 0
 Stopped: 4
Images: 47
Server Version: 18.09.3
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: active
 NodeID: [redacted]
 Is Manager: true
 ClusterID: [redacted]
 Managers: 1
 Nodes: 1
 Default Address Pool: 10.0.0.0/8
 SubnetSize: 24
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 10
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: [redacted]
 Manager Addresses:
  [redacted]:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: e6b3f5632f50dbc4e9cb6288d911bf4f5e95b18e
runc version: 6635b4f0c6af3810594d2770f662f34ddc15b40d
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.0-8-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.801GiB
Name: server
ID: UIPI:DIKQ:34LS:Z65A:3VHZ:S6VL:XHME:Z6AC:53NG:3WHR:CCMM:QHW7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: [redacted]
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

WARNING: No swap limit support
```

- Output of `docker-compose version` (if applicable):

```
docker-compose version 1.23.2, build 1110ad01
docker-py version: 3.6.0
CPython version: 3.6.7
OpenSSL version: OpenSSL 1.1.0f  25 May 2017
```

**Comment by tompizmor** [comment]
Hi @nicokaiser,

> When setting the log level to TRACE8, the log reveals that this might have something to to with nami trying to chown all files all over again (and doing so in a very slow way, as there are not that many files in the newly created volume). This should not happen when the .initialized file is found.

I want to recall that this was an issue on older versions of the WordPress container but it should have been already fixed. 
Can you let me know which WordPress tag are you using? In case it is `latest` make sure you update your local copy using `docker-compose pull`.

**Comment by nicokaiser** [comment]
I am using the latest tag (of bitnami/wordpress-nginx container, but I experienced this also for bitnami/wordpress:latest).

**Comment by tompizmor** [comment]
I can see the chown/chmod commands you mention. However in my tests it only takes a couple of seconds.

Can you launch one testing container and measure the time of a recursive chwon in the `/opt/bitnami/wordpress` directory? Did you always test the Wordpress container in the same machine? If so, would be possible for you to try on a different host?

```
$ docker run -it --rm --name wordpress-test bitnami/wordpress:latest bash

Welcome to the Bitnami wordpress container
Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-wordpress
Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-wordpress/issues

root@71220e1ca0f5:/# time chown -R bitnami:bitnami /opt/bitnami/wordpress/

real	0m12.701s
user	0m0.016s
sys	0m0.780s
```

Note it took 12 secods in my test.

**Log output**
```
wordpress_1  | nami    INFO  Initializing wordpress
wordpress_1  | wordpre TRACE [normalize] File normalization: tmp -> /opt/bitnami/wordpress/tmp
wordpress_1  | wordpre TRACE [mkdir] File normalization: /opt/bitnami/wordpress/tmp -> /opt/bitnami/wordpress/tmp
wordpress_1  | wordpre TRACE [exists] File normalization: /opt/bitnami/apache/conf -> /opt/bitnami/apache/conf
...
wordpress_1  | wordpre TRACE [chown] File normalization: /opt/bitnami/wordpress/wp-signup.php -> /opt/bitnami/wordpress/wp-signup.php
wordpress_1  | wordpre TRACE [configurePermissions] File to chmod: /opt/bitnami/wordpress/wp-signup.php
wordpress_1  | wordpre TRACE [chmod] File normalization: /opt/bitnami/wordpress/wp-signup.php -> /opt/bitnami/wordpress/wp-signup.php
wordpress_1  | wordpre TRACE [isLink] File normalization: /opt/bitnami/wordpress/wp-trackback.php -> /opt/bitnami/wordpress/wp-trackback.php
wordpress_1  | wordpre TRACE [isDirectory] File normalization: /opt/bitnami/wordpress/wp-trackback.php -> /opt/bitnami/wordpress/wp-trackback.php
wordpress_1  | wordpre TRACE [isLink] File normalization: /opt/bitnami/wordpress/wp-trackback.php -> /opt/bitnami/wordpress/wp-trackback.php
wordpress_1  | wordpre TRACE [configurePermissions] File to chown: /opt/bitnami/wordpress/wp-trackback.php
wordpress_1  | wordpre TRACE [chown] File normalization: /opt/bitnami/wordpress/wp-trackback.php -> /opt/bitnami/wordpress/wp-trackback.php
wordpress_1  | wordpre TRACE [configurePermissions] File to chmod: /opt/bitnami/wordpress/wp-trackback.php
wordpress_1  | wordpre TRACE [chmod] File normalization: /opt/bitnami/wordpress/wp-trackback.php -> /opt/bitnami/wordpress/wp-trackback.php
wordpress_1  | wordpre TRACE [isLink] File normalization: /opt/bitnami/wordpress/xmlrpc.php -> /opt/bitnami/wordpress/xmlrpc.php
wordpress_1  | wordpre TRACE [isDirectory] File normalization: /opt/bitnami/wordpress/xmlrpc.php -> /opt/bitnami/wordpress/xmlrpc.php
wordpress_1  | wordpre TRACE [isLink] File normalization: /opt/bitnami/wordpress/xmlrpc.php -> /opt/bitnami/wordpress/xmlrpc.php
wordpress_1  | wordpre TRACE [configurePermissions] File to chown: /opt/bitnami/wordpress/xmlrpc.php
wordpress_1  | wordpre TRACE [chown] File normalization: /opt/bitnami/wordpress/xmlrpc.php -> /opt/bitnami/wordpress/xmlrpc.php
wordpress_1  | wordpre TRACE [configurePermissions] File to chmod: /opt/bitnami/wordpress/xmlrpc.php
wordpress_1  | wordpre TRACE [chmod] File normalization: /opt/bitnami/wordpress/xmlrpc.php -> /opt/bitnami/wordpress/xmlrpc.php
wordpress_1  | wordpre TRACE [exists] File normalization: /bitnami/wordpress/.initialized -> /bitnami/wordpress/.initialized
wordpress_1  | wordpre INFO  WordPress has been already initialized, restoring...
```

**Comment by nicokaiser** [comment]
Thanks for your comment! I tested the chown call on two machines:

An EC2 test instance (t2.micro using a gp2 EBS volume (SSD)):

```
root@d7c38f2990cf:/# time chown -R bitnami:bitnami /opt/bitnami/wordpress/

real	0m53.992s
user	0m0.017s
sys	0m0.675s
```

My production machine (AMD Opteron 2220, 2.8 GHz, via VMware vSphere, using a hardware RAID of spinning disks (no SSD)):

```
root@87795f69d3c2:/# time chown -R bitnami:bitnami /opt/bitnami/wordpress/

real	2m52.988s
user	0m0.036s
sys	0m1.772s
```

One minute is already slow, but three minutes is even worse. Maybe this would be way faster if the chown was run at container creation (Dockerfile)? Performing 6.000 chown operations on the layered Docker filesystem is not very efficient...

**Comment by nicokaiser** [comment]
In https://github.com/docker/for-linux/issues/388 this is also mentioned. Recursive chown on CoW filesystems are slow sometimes...

**Comment by tompizmor** [comment]
Moving some chmod/chwon commands to the Dockerfile is something that we have already done for other apps ( [magento for example](https://github.com/bitnami/bitnami-docker-magento/blob/master/2/debian-9/Dockerfile#L15) ), so we can consider it for wordpress too. 

However, I still think that a chown command should not be that slow. I checked the issue you shared and it seems a lot of people is affected by this issue but it seems there isn't any workaround 100% valid.

**Comment by nicokaiser** [comment]
Preparing the container's contents should be done in the Dockerfile and chmod/chown for a container-directory does not need to run at every start as it will not change during runtime (unless a volume is mounted at /opt/bitnami/wordpress, but then the user knows what they are doing, so this case can safely be ignored).

Running chmod over 6.000 files on a CoW filesystem is not very efficient (see the issue). When using fast SSDs, this may not be a big problem, but it is not very elegant and can be avoided.

**Comment by nicokaiser** [comment]
I tried to create a PR with these lines (chmod/chown in Dockerfile), however I think this does not prevent `nami_initialize` running through all files again, so this is not a real solution until the WordPress initialization stops doing this on every startup.

Don't know if this is feasable, as the wordpress packages are used elsewhere (where this acutally makes sense) I suppose?

**Comment by tompizmor** [comment]
You are right! The chown/chmod commands will be executed as part as the nami module initialization unless we modify that on our side.

I am afraid right now that code is not publicly available. However, you can see it if you exec into the container and check the file `/root/.nami/components/com.bitnami.wordpress/main.js`. 

You could even copy that file outside the container with `docker cp`, edit it on your host, and then copy again in your custom Dockerfile.

**Comment by nicokaiser** [comment]
So maybe a solution would be if 

1. the Dockerfile includes chown/chmod commands for /opt/bitnami/wordpress
2. the Bitnami wordpress package only execues chown/chmod for files with the wrong permissions.

This way unnecessary CoW operations are avoided which makes container startup much faster.

**Comment by tompizmor** [comment]
Agree. I have created a task in our internal system and we will take it into account for future WordPress releases.

**Comment by nicokaiser** [comment]
Sounds good, thanks!

**Comment by nicokaiser** [comment]
Any news on this?

**Comment by tompizmor** [comment]
I am afraid we couldn't find the time to work on this yet.

**Comment by hameno** [comment]
This is very annoying for us as well...

**Comment by tompizmor** [comment]
@hameno can you let us know how was your Kubernetes cluster provisioned? 

I would like to understand why is slow in some clusters and fast in others.

**Comment by hameno** [comment]
Standard Azure AKS, initially with a Azure Files mount, but that was unusable because of the bad perf of that storage class, now with a Rook NFS share on premium storage

**Comment by nicokaiser** [comment]
Like I said, thousands (!) of unnecessary chown operations at every container start is not very elegant. This could be avoided completely, but needs adjustments in the Bitnami WordPress package. I cannot make a PR there because it is not public...

**Comment by miguelaeh** [comment]
Hi, 
be patient please, we will work on that when we find the time.
Thank you, and thank you for reporting this issue. We will update this thread when we have news on that.

**Comment by markusschuettler** [comment]
Encountered the same issue with a HostPath pv on a single node k3s environment.

**Comment by tompizmor** [comment]
Thanks for the info @markusschuettler.

**Comment by markusschuettler** [comment]
Short update in case it helps anyone:
I managed to circumvent the issue by upgrading the kernel of node host (to >=4.19) and enabling ["Metadata only copy up" for OverlayFS](https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt). This drastically increased the start up time of the pod.

**Comment by tompizmor** [comment]
Great news! Thanks for sharing

**Comment by Chematronix** [comment]
How can I confirm I'm being affected by this?  Can't get past the following even after 30 min:

```
namiwordpress | apache  TRACE [runProgram] Executing internal command: '/bin/sh' ["-c","/opt/bitnami/apache/bin/httpd -f /opt/bitnami/apache/conf/httpd.conf &"]
namiwordpress | apache  TRACE [runProgram] RESULT: {"code":0,"stderr":"","stdout":""}
namiwordpress | apache  TRACE [normalize] File normalization: tmp -> /opt/bitnami/apache/tmp
namiwordpress | apache  TRACE [normalize] File normalization: /opt/bitnami/apache/tmp/httpd.pid -> /opt/bitnami/apache/tmp/httpd.pid
namiwordpress | apache  TRACE [normalize] File normalization: tmp -> /opt/bitnami/apache/tmp
namiwordpress | apache  TRACE [normalize] File normalization: /opt/bitnami/apache/tmp/httpd.pid -> /opt/bitnami/apache/tmp/httpd.pid
namiwordpress | apache  TRACE [normalize] File normalization: tmp -> /opt/bitnami/apache/tmp
namiwordpress | apache  TRACE [normalize] File normalization: /opt/bitnami/apache/tmp/httpd.pid -> /opt/bitnami/apache/tmp/httpd.pid
namiwordpress | apache  TRACE [normalize] File normalization: tmp -> /opt/bitnami/apache/tmp
namiwordpress | apache  TRACE [normalize] File normalization: /opt/bitnami/apache/tmp/httpd.pid -> /opt/bitnami/apache/tmp/httpd.pid
namiwordpress | apache  TRACE [normalize] File normalization: tmp -> /opt/bitnami/apache/tmp
namiwordpress | apache  TRACE [normalize] File normalization: /opt/bitnami/apache/tmp/httpd.pid -> /opt/bitnami/apache/tmp/httpd.pid
namiwordpress | apache  TRACE [normalize] File normalization: tmp -> /opt/bitnami/apache/tmp
namiwordpress | apache  TRACE [normalize] File normalization: /opt/bitnami/apache/tmp/httpd.pid -> /opt/bitnami/apache/tmp/httpd.pid
namiwordpress | apache  TRACE [normalize] File normalization: tmp -> /opt/bitnami/apache/tmp
namiwordpress | apache  TRACE [normalize] File normalization: /opt/bitnami/apache/tmp/httpd.pid -> /opt/bitnami/apache/tmp/httpd.pid
namiwordpress | apache  TRACE code: 0
namiwordpress | apache  TRACE stdout:
namiwordpress |
namiwordpress | apache  TRACE stderr:
namiwordpress |
namiwordpress | apache  TRACE [normalize] File normalization: logs -> /opt/bitnami/apache/logs
namiwordpress | apache  TRACE [normalize] File normalization: /opt/bitnami/apache/logs/access_log -> /opt/bitnami/apache/logs/access_log
namiwordpress | apache  TRACE [normalize] File normalization: logs -> /opt/bitnami/apache/logs
namiwordpress | apache  TRACE [normalize] File normalization: /opt/bitnami/apache/logs/access_log -> /opt/bitnami/apache/logs/access_log
namiwordpress |

```

**Comment by javsalgar** [comment]
Could you let us know about the file system you are using?

**Comment by Chematronix** [comment]
> Could you let us know about the file system you are using?

NTFS on a spinny drive.  Things got moving after a system restart, but still very slow to init, so I switched to another image.

**Comment by javsalgar** [comment]
I presume that it has to do with the chown, it may not work well with that underlying system. We will update the ticket as soon as we have more information about improving the chown/chmod

**Comment by woodcockjosh** [comment]
I am using local SSD and I also am experiencing extremely slow initialization.

**Comment by juan131** [comment]
Hi @woodcockjosh 

Could you run the commands below and share the output? That will increase the verbosity of the logs and we'll have more information to debug what's going on:

```console
$ cat > values.yaml <<'EOF'
extraEnv:
  - name: NAMI_LOG_LEVEL
    value: "trace"
EOF
$ helm install -f values.yaml my-wordpress stable/wordpress
```

It's important to understand what specific action is taking so long to be completed.

**Comment by nicokaiser** [comment]
Files (many of them) are being chmod/chown'ed on every startup. This is highly inefficient as this should be done on image creation, and on spinning disks and/or slow storage systems it is really slow due to the copy-on-write nature of Docker containers.

**Comment by woodcockjosh** [comment]
You can't do a chmod in docker build for all the files because the files being chmod are on a volume. I would suggest that this be done inside of a 1 time setup step rather than every time and use bash rather than node for performance.

**Comment by juan131** [comment]
Hi @woodcockjosh 

At some point will move the whole logic from node to bash on every web application (as we did for "infrastructure" applications such as MariaDB or PostgreSQL).

We did a workaround for other solutions (such as Magento) to adapt the permissions using chmod/chown before moving the files to the volume as you can check in the link below:

- https://github.com/bitnami/bitnami-docker-magento/blob/master/2/debian-9/Dockerfile#L12

We could sth similar on WordPress. However, this is not compatible with running the WP container with a non-root user.

**Comment by woodcockjosh** [comment]
If you use gosu you can update the file permissions before you switch to non-root user in the entrypoint script.

**Comment by juan131** [comment]
Hi @nicokaiser 

> If you use gosu you can update the file permissions before you switch to non-root user in the entrypoint script.

Could you please extend that? I don't understand what you're proposing.
When running the container as non-root, the **entrypoint.sh** will be executed as non-root, and the UID of the non-root user is arbitrary. You don't know in advance what UID to use to adapt the permissions.

**Comment by woodcockjosh** [comment]
If you want to go about adjusting permissions on the fly you need to run the image as root eg `USER root` and then after the permissions are updated use `gosu` to switch the user to the wordpress user. The effect is the same as putting `USER wordpress` in the Dockerfile. Many other images are doing this for the very same reason. I do think you need to investigate the root cause of the file permissions issue and determine whether or not updating the permissions on the fly is really necessary. You could for example create a feature to detect when the file permissions are inaccurate in the administrative panel and ask the user to initiate an update manually. For the most part when a brand new installation of wordpress on k8s is occurring, it is also a brand new instance of wordpress, meaning that files are not being migrated. Therefore you could also update the permissions during a setup/installation phase instead.

**Comment by javsalgar** [comment]
Hi,

Thanks for the insight. I do not think that using `gosu` would work on platforms like Openshift, where the user is random.

I think that the best way to avoid the chown/chmod operations would be to have a non-root container. That should prevent us this kind of situations and have something like @juan131 shared.

We will update the ticket when we have more information.

**Comment by marcosbc** [comment]
Hi, AFAIK this issue was already fixed in the `5.4.1-debian-10-r18` release of the Bitnami WordPress container image. I'm closing this issue, but if you continue to find performance issues please let us know.
